pkg/adapter/config/encryptionAdapter.go

```diff
@@ -2,29 +2,132 @@ package config
+	return encryptedEntries, errors.Join(encryptionErrors...)
+}
+
+func (p PublicKeyConfigEncryptionAdapter) Decrypt(
+	ctx context.Context,
+	name common.SimpleDoguName,
+	encryptedValue common.EncryptedDoguConfigValue) (common.SensitiveDoguConfigValue, error) {
+	privateKeySecret, err := p.secrets.Get(ctx, string(name)+"-private", metav1.GetOptions{})
+	if err != nil {
+		return "", domainservice.NewNotFoundError(err, "could not get private key")
+	}
+	privateKey := privateKeySecret.Data["private.pem"]
+	keyPair, err := getKeyPairFromPrivateKey(privateKey, p.registry)
+	if err != nil {
+		return "", domainservice.NewInternalError(err, "could not get key pair for dogu %v", string(name))
+	}
+	decryptedValue, err := keyPair.Private().Decrypt(string(encryptedValue))
+	if err != nil {
+		return "", domainservice.NewInternalError(err, "could not decrypt encrypted value")
+	}
+	return common.SensitiveDoguConfigValue(decryptedValue), nil
+}
+
+func getKeyPairFromPrivateKey(privateKey []byte, registry etcdRegistry) (*keys.KeyPair, error) {
+	keyProvider, err := resource.GetKeyProvider(registry)
+	if err != nil {
+		return &keys.KeyPair{}, domainservice.NewNotFoundError(err, "could not get key provider")
+	}
+	keyPair, err := keyProvider.FromPrivateKey(privateKey)
+	if err != nil {
+		return &keys.KeyPair{}, domainservice.NewInternalError(err, "could not get key pair")
+	}
+	return keyPair, nil
+}
+
+func (p PublicKeyConfigEncryptionAdapter) DecryptAll(ctx context.Context, entries map[common.SensitiveDoguConfigKey]common.EncryptedDoguConfigValue) (map[common.SensitiveDoguConfigKey]common.SensitiveDoguConfigValue, error) {
+	decryptedEntries := map[common.SensitiveDoguConfigKey]common.SensitiveDoguConfigValue{}
+	var decryptionErrors []error
+	keypairs := map[string]*keys.KeyPair{}
+	for configKey, configValue := range entries {
+		doguname := string(configKey.DoguName)
+		keyPair, privateKnown := keypairs[doguname]
+		if !privateKnown {
+			privateKeySecret, err := p.secrets.Get(ctx, doguname+"-private", metav1.GetOptions{})
+			if err != nil {
+				decryptionErrors = append(decryptionErrors, domainservice.NewNotFoundError(err, "could not get private key for dogu %v", doguname))
+				continue
+			} else {
+				privateKey := privateKeySecret.Data["private.pem"]
+				keyPairFromRegistry, err := getKeyPairFromPrivateKey(privateKey, p.registry)
+				if err != nil {
+					decryptionErrors = append(decryptionErrors, domainservice.NewNotFoundError(err, "could not get key pair for dogu %v", doguname))
+					continue
+				} else {
+					keyPair = keyPairFromRegistry
+				}
+			}
+		}
+		decryptedValue, err := keyPair.Private().Decrypt(string(configValue))
+		if err != nil {
+			decryptionErrors = append(decryptionErrors, domainservice.NewInternalError(err, "could not decrypt encrypted value for dogu %v", doguname))
+		}
+		decryptedEntries[configKey] = common.SensitiveDoguConfigValue(decryptedValue)
+	}
+	return decryptedEntries, errors.Join(decryptionErrors...)
```

Comment:
Also consider here to load keyPairs beforehand with a helper function. You can also introduce a helper function `decryptWithKey(keyPair, encryptedValue) (sensitiveValue, err)` (or something similar) to reuse more code between `decrypt` and `decryptAll`.